stan/math/prim/fun/mdivide_left_ldlt.hpp

```diff
@@ -13,29 +13,31 @@ namespace math {
 /**
  * Returns the solution of the system Ax=b given an LDLT_factor of A
  *
- * @tparam T1 type of elements in the LDLT_factor
- * @tparam T2 type of elements in right-hand side matrix or vector
- * @tparam R1 number of rows in the LDLT_factor, can be Eigen::Dynamic
- * @tparam C1 number of columns in the LDLT_factor, can be Eigen::Dynamic
- * @tparam R2 number of rows in the right-hand side matrix, can be
- *         Eigen::Dynamic
- * @tparam C2 number of columns in the right-hand side matrix, can be
- *         Eigen::Dynamic
+ * @tparam T type of elements in the LDLT_factor
+ * @tparam R number of rows in the LDLT_factor, can be Eigen::Dynamic
+ * @tparam C number of columns in the LDLT_factor, can be Eigen::Dynamic
+ * @tparam EigMat type of the matrix
  *
  * @param A LDLT_factor
  * @param b Right hand side matrix or vector.
  * @return x = A^-1 b, solution of the linear system.
  * @throws std::domain_error if rows of b don't match the size of A.
  */
-template <int R1, int C1, int R2, int C2, typename T1, typename T2>
-inline Eigen::Matrix<return_type_t<T1, T2>, R1, C2> mdivide_left_ldlt(
-    const LDLT_factor<T1, R1, C1> &A, const Eigen::Matrix<T2, R2, C2> &b) {
+template <int R, int C, typename T, typename EigMat,
+          require_eigen_t<EigMat>* = nullptr,
+          require_all_not_st_var<T, EigMat>* = nullptr,
+          require_not_t<conjunction<std::is_arithmetic<T>,
+                                    is_fvar<value_type_t<EigMat>>>>* = nullptr>
+inline Eigen::Matrix<return_type_t<T, EigMat>, R, EigMat::ColsAtCompileTime>
+mdivide_left_ldlt(const LDLT_factor<T, R, C>& A, const EigMat& b) {
   check_multiplicable("mdivide_left_ldlt", "A", A, "b", b);
   if (A.cols() == 0) {
     return {0, b.cols()};
   }
 
-  return A.solve(Eigen::Matrix<return_type_t<T1, T2>, R2, C2>(b));
+  return A.solve(
+      Eigen::Matrix<return_type_t<T, EigMat>, EigMat::RowsAtCompileTime,
+                    EigMat::ColsAtCompileTime>(b));
```

Comment:
I dislike using auto for return types, because it hides what the return type is. That I have to search for return statements to determine return type. In this case that is not enough. Next I have to search what `A.solve(ret_mat(b)).eval();` returns or what `ret_mat` is. 

In general code if more often read than changed. That is why I think we should prefere readability to ease of implementation.

About RVO: In this functions we need to construct return matrix anyway. We gain nothing performance-wise if we construct it explicitely and can do RVO. Instead what I wrote does these conversions implicitely.

> Also having a bit of a hard time following the requires for this function.

That last require is needed not to use this implementation for what we have a fwd implementation. It lools confusing because we only have fwd implementation for the case of ` LDLT_factor == double` and `EigMat = fvar`. I checked it again and it is correct. I did simplify it the way you suggested.

Reply:
I'm fine with leaving it as is

> About RVO: In this functions we need to construct return matrix anyway. We gain nothing performance-wise if we construct it explicitely and can do RVO. Instead what I wrote does these conversions implicitely.

This was my bad because I forgot that if you have implicit constructors available and returning temps RVO can still happen. Yes the matrix is constructed in the function either way but RVO let's the memory it's assigned to be at the return callee and not copied (or moved via NRVO) right?

Reply:
> I dislike using auto for return types

On this though, I prefer them for some functions where we just use eigen functions since a lot of the time we can reuse that code for the sparse matrix stuff I'm doing right now. i.e. for add, mul, etc. those functions can currently take in and return sparse matrices

Reply:
Though a lot of these wouldn't work for sparse matrices so it's not an issue here